src/colmap/estimators/covariance.cc

```diff
@@ -0,0 +1,453 @@
+  return true;
+}
+
+bool BundleAdjustmentCovarianceEstimatorCeresBackend::Compute() {
+  ceres::Covariance::Options options;
+  ceres::Covariance covariance_computer(options);
+  std::vector<const double*> parameter_blocks;
+  parameter_blocks.insert(
+      parameter_blocks.end(), pose_blocks_.begin(), pose_blocks_.end());
+  if (!covariance_computer.Compute(parameter_blocks, problem_)) return false;
+  Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> covs(
+      num_params_poses_, num_params_poses_);
+  covariance_computer.GetCovarianceMatrixInTangentSpace(parameter_blocks,
+                                                        covs.data());
+  cov_poses_ = covs;
+  return true;
+}
+
+void BundleAdjustmentCovarianceEstimator::SetLambda(double lambda) {
+  lambda_ = lambda;
+}
+
+void BundleAdjustmentCovarianceEstimator::ComputeSchurComplement() {
+  // Evaluate jacobian
+  LOG(INFO) << "Evaluate jacobian matrix";
+  ceres::Problem::EvaluateOptions eval_options;
+  eval_options.parameter_blocks.clear();
+  for (const double* block : pose_blocks_) {
+    eval_options.parameter_blocks.push_back(const_cast<double*>(block));
+  }
+  for (const double* block : other_variables_blocks_) {
+    eval_options.parameter_blocks.push_back(const_cast<double*>(block));
+  }
+  for (const double* block : point_blocks_) {
+    eval_options.parameter_blocks.push_back(const_cast<double*>(block));
+  }
+  ceres::CRSMatrix J_full_crs;
+  problem_->Evaluate(eval_options, nullptr, nullptr, nullptr, &J_full_crs);
+  int num_residuals = J_full_crs.num_rows;
+  int num_params = J_full_crs.num_cols;
+  Eigen::Map<Eigen::SparseMatrix<double, Eigen::RowMajor>> J_full(
+      J_full_crs.num_rows,
+      J_full_crs.num_cols,
+      J_full_crs.values.size(),
+      J_full_crs.rows.data(),
+      J_full_crs.cols.data(),
+      J_full_crs.values.data());
+
+  // Step 3: Schur elimination on points
+  LOG(INFO) << "Schur elimination on points";
+  Eigen::SparseMatrix<double> J_c =
+      J_full.block(0, 0, num_residuals, num_params - num_params_points_);
+  Eigen::SparseMatrix<double> J_p = J_full.block(
+      0, num_params - num_params_points_, num_residuals, num_params_points_);
+  Eigen::SparseMatrix<double> H_cc = J_c.transpose() * J_c;
+  Eigen::SparseMatrix<double> H_cp = J_c.transpose() * J_p;
+  Eigen::SparseMatrix<double> H_pc = H_cp.transpose();
+  Eigen::SparseMatrix<double> H_pp = J_p.transpose() * J_p;
+  // in-place computation of H_pp_inv
+  Eigen::SparseMatrix<double>& H_pp_inv = H_pp;  // reference
+  int counter_p = 0;
+  for (const double* block : point_blocks_) {
+    int num_params_point = ParameterBlockTangentSize(problem_, block);
+    Eigen::SparseMatrix<double> subMatrix_sparse =
+        H_pp.block(counter_p, counter_p, num_params_point, num_params_point);
+    Eigen::MatrixXd subMatrix = subMatrix_sparse;
+    subMatrix +=
+        lambda_ * Eigen::MatrixXd::Identity(subMatrix.rows(), subMatrix.cols());
```

Comment:
Using pseudo inverse here is potentially incorrect for covariance computation. It is setting the reciprocal of all the near-zero values to be zero, which should be large values instead. This can potentially result in underestimation of the covariance. Instead, damping the matrix is generally used for LM optimization and is considered a general practice to handle the singularity.

Reply:
OK maybe I am the one who is confused here. This might be valid. Then I have a cleaner proposal: how about we just dropping all the damping factors and apply MP inverse for both here and the 3x3 inversion at the point blocks? : )

Reply:
You are right: in the direction that is unconstrained, the pseudo-inverse would yield a zero variance while the damping simply caps it to a maximum value. Then how about replacing the damping by an SVD-based inverse, where we clip the reciprocal of the singular values to a maximum threshold? (~1/damping)

Reply:
Regarding the damping of the other parameters: how often is this needed? aren't they most often sufficiently constrained?

Reply:
Depending on the camera model and scene geometry, the estimation of intrinsics can often be poorly conditioned even for many residuals in the images.

Reply:
Although I agree that using SVD to clamp the singular values may be a slightly better solution, it is hard to make efficient. Now we are using Cholesky which is much efficient than SVD. I have changed the default value of lambda to 1e-8 and believe that this should not be a problem numerically.

Reply:
For the 3x3 point blocks, you are currently using LU factorization. For 3x3 matrices, I doubt that there is much difference in efficiency between this and SVD. This can also be trivially parallelized over all points with a thread pool. But eventually I'm unsure whether damping vs s-value clamping really matters.

Reply:
I don't have any intuition on whether damping or SVD is better. I'll leave the decision to the two of you.